### PR TITLE
Launch Yeoman Generator in separate thread

### PR DESCRIPTION
**Pull request by vitaliy-guliy**
As VS Code extension generator goes not take in attention an option specifying the directory for generated projects, it always generate a project into current working directory.
The idea was to run Yeoman generator in separate node instance and change current working directory to `/projects`. 

Fixes issue https://github.com/eclipse/che/issues/15443

Subtask of https://github.com/eclipse/che/issues/14228

### CODE REVIEW

File: src/theia-yeoman-backend-plugin.ts [hunk 1]
```diff
@@ -0,0 +1,38 @@
+/*********************************************************************
+* Copyright (c) 2018 Red Hat, Inc.
```
Comment (benoitf) [inline src/theia-yeoman-backend-plugin.ts]: as you're doing a refactoring, copyright date needs to be updated

File: src/yeoman-process-adapter.ts [hunk 2]
```diff
@@ -0,0 +1,221 @@
+/*********************************************************************
+* Copyright (c) 2018 Red Hat, Inc.
```
Comment (benoitf) [inline src/yeoman-process-adapter.ts]: please update

File: src/yeoman-process-adapter.ts [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+* SPDX-License-Identifier: EPL-2.0
+**********************************************************************/
+
+import * as theia from '@theia/plugin';
+import { ChildProcess } from 'child_process';
+
+const path = require('path');
+const fork = require('child_process').fork;
+
+export class YeomanProcessAdapter {
```
Comment (benoitf) [inline src/yeoman-process-adapter.ts]: please comment

File: src/yeoman-process-adapter.ts [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+            return;
+        }
+
+        this.outputChannel.show(true);
+
+        // grab first workspace
+        const currentWorkspace = workspaceFolders[0];
+
+        // start Yeoman generator in separate node instance
+        const scriptPath = this.extensionPath + '/lib/yeoman-process.js';
```
Comment (benoitf) [inline src/yeoman-process-adapter.ts]: we should use Path.resolve no ?

File: src/yeoman-process-adapter.ts [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+            cwd: currentWorkspace.uri.path,
+            stdio: ['pipe', process.stdout, process.stderr, 'ipc']
+        };
+
+        const child = fork(process, params, options);
+
+        child.on('message', this.onMessage);
+
+        child.stdout.on('data', (data: any) => {
+            console.log(`## stdout: ${data}`);
```
Comment (benoitf) [inline src/yeoman-process-adapter.ts]: why do we need to output to the console stdout or stderr  ?

File: src/yeoman-process.ts [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+* which is available at https://www.eclipse.org/legal/epl-2.0/
+*
+* SPDX-License-Identifier: EPL-2.0
+**********************************************************************/
+
+import * as yeoman from 'yeoman-environment';
+
+const proc = process;
+
+export class YeomanProcess {
```
Comment (benoitf) [inline src/yeoman-process.ts]: please comment ?

**Comment by benoitf** [comment]
Hello,
in the current code, it's doing:
```
    const currentWorkspace = workspaceFolders[0];	
    yeomanOpts.cwd = currentWorkspace.uri.path;	
    const yeomanEnvironment = yeoman.createEnv(undefined, yeomanOpts, new TheiaYeomanAdapter());
```

so AFAIK it launches yeoman from the workspace folder which should be `/projects` no ?

**Comment by vitaliy-guliy** [comment]
> Hello,
> in the current code (before your pull request), it's doing:
> 
> ```
>     const currentWorkspace = workspaceFolders[0];	
>     yeomanOpts.cwd = currentWorkspace.uri.path;	
>     const yeomanEnvironment = yeoman.createEnv(undefined, yeomanOpts, new TheiaYeomanAdapter());
> ```
> 
> so AFAIK it launches yeoman from the workspace folder which should be `/projects` no ?
> why do we need a fork process ?

VS Code extension generator does not take into account `cwd` configuration property and always generate the project into current working directory, which is `/home/theia`. That's why we need to launch a separate process and change the working directory on `/projects`.

In the current code, VS Code yeoman generator is running in current node instance, not in separate.

File: src/yeoman-process-adapter.ts (hunk 5 above)
Comment (vitaliy-guliy) [reply src/yeoman-process-adapter.ts]: It helps us to understand the reason when something goes wrong.
Once it helped me to catch an error when running the generator.

File: src/theia-yeoman-backend-plugin.ts (hunk 1 above)
Comment (vitaliy-guliy) [reply src/theia-yeoman-backend-plugin.ts]: Done

File: src/yeoman-process-adapter.ts (hunk 2 above)
Comment (vitaliy-guliy) [reply src/yeoman-process-adapter.ts]: Done

File: src/yeoman-process-adapter.ts (hunk 3 above)
Comment (vitaliy-guliy) [reply src/yeoman-process-adapter.ts]: Done

File: src/yeoman-process-adapter.ts (hunk 4 above)
Comment (vitaliy-guliy) [reply src/yeoman-process-adapter.ts]: Done

File: src/yeoman-process-adapter.ts (hunk 3 above)
Comment (vitaliy-guliy) [reply src/yeoman-process-adapter.ts]: Strange, GitHub does not display the changes but they are present
https://github.com/vitaliy-guliy/theia-yeoman-plugin/blob/separate-thread/src/yeoman-process-adapter.ts

File: src/yeoman-process.ts (hunk 6 above)
Comment (vitaliy-guliy) [reply src/yeoman-process.ts]: Done

File: src/yeoman-process-adapter.ts (hunk 5 above)
Comment (benoitf) [reply src/yeoman-process-adapter.ts]: why it's not part of the output on the UI side then ?

IMHO we shouldn't pollute that console by default (= not when debugging)

Comment (vitaliy-guliy) [reply src/yeoman-process-adapter.ts]: UI displays only the messages related to creating the extension.
The messages form stdout / stderr are printed to Theia backend console, which is not visible in Theia.
It's easy to ignore them, but should we?

Comment (vitaliy-guliy) [reply src/yeoman-process-adapter.ts]: I have reworked it a bit. All the output has been printed to the UI. Looks nice.

**Comment by vitaliy-guliy** [comment]
I tried to do it before. You could scroll up to see the commit where I returned it back to 'theia:plugin'.
On the local machine 'theia-plugin' does not work also
```
yarn run v1.17.3
$ yarn run format-code && yarn run compile && yarn run tslint-fix && theia-plugin pack
$ tsfmt -r
$ tsc
$ tslint --fix --project .
/bin/sh: 1: theia-plugin: not found
error Command failed with exit code 127.
```

**Comment by vitaliy-guliy** [comment]
@benoitf could you merge this PR? It seems only you have right do it.

**Comment by benoitf** [comment]
I can merge but we need to fix first the ci issue else it won't be built and then it won't be available

**Pull request merged**